accession-commons-monotonic-generator-jpa/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/jpa/monotonic/service/ContiguousIdBlockService.java

```diff
@@ -74,12 +76,22 @@ public BlockParameters getBlockParameters(String categoryId) {
         return categoryBlockInitializations.get(categoryId);
     }
 
-    @Transactional(readOnly = true)
-    public List<ContiguousIdBlock> getUncompletedBlocksByCategoryIdAndApplicationInstanceIdOrderByEndAsc(
+    @Transactional(isolation = Isolation.SERIALIZABLE)
+    public List<ContiguousIdBlock> reserveUncompletedBlocksByCategoryIdAndApplicationInstanceIdOrderByEndAsc(
             String categoryId, String applicationInstanceId) {
         try (Stream<ContiguousIdBlock> reservedBlocksOfThisInstance = repository
                 .findAllByCategoryIdAndApplicationInstanceIdOrderByLastValueAsc(categoryId, applicationInstanceId)) {
-            return reservedBlocksOfThisInstance.filter(ContiguousIdBlock::isNotFull).collect(Collectors.toList());
+            List<ContiguousIdBlock> blocksList = reservedBlocksOfThisInstance
+                    .filter(block -> block.isNotFull() && block.isNotReserved()).collect(Collectors.toList());
+
+            blocksList.stream().forEach(block -> block.markAsReserved());
+            save(blocksList);
+
+            return blocksList;
```

Comment:
If I'm not mistaken, this is the core of the process safe block reservation process. I feel like we should describe in a comment why/how this works.

Reply:
Maybe it makes more sense to describe this at the level of the class rather than this method.

Reply:
Added description on top of the class in the form of Javadoc to explain the behavior